### PR TITLE
Finish migrating tileset layers to classes

### PR DESCRIPTION
**Pull request by lmoureaux**
This concludes a refactoring of the tileset layer code initiated three years ago in #450. It doesn't mean that the tileset code is now perfect, but it's easier to work with. No user-facing change is expected.

Warning: **very large** PR. I think we need to discuss the merging strategy:
* Split in many small PRs for a manageable code review or only test the final product? We could simply try all known tilesets on a few maps and check that the PR introduces no difference (save map to image -> compare images?).
* Do we wait for the 3.2 branch to be created?

The idea is to remove the large `switch`-based function that used to dispatch tile drawing and replace it with a series of `layer_xxx` classes, each of them responsible for one type of mapview layers. Dispatching to the correct class is done through inheritance. Since each class cares about a single layer, they can be small, making them easier to understand. It is also guaranteed that no other code fiddles with the class data. 

This PR removes about 2000 lines from `tilespec.cpp`. What remains in that file is mainly the tileset loading logic including functions to cut out sprites from the tileset pngs, and many getter functions to query tileset properties.

Migrating to classes required at least minor changes to the code to update sprite variable names. In many cases I also modified the logic to be easier to follow. An extreme example is `layer_roads` which was basically rewritten from scratch.

### CODE REVIEW

**Comment by jwrober** [comment]
FWIW, I am not in favor of breaking this PR out into many. What I do think we might want to do is advertise this on Discord to see if other players that are not afraid of running pre-release code to put this through some varying tests. @blabber @hugomflavio come to mind.

**Comment by jwrober** [comment]
I'd like to see if there is a way to reduce the `stdout` output when running the client from the command line.

**Comment by jwrober** [comment]
Sprite artifact in city dialog should be removed
image

**Comment by lmoureaux** [comment]
`stdout` and wall sprite done. There is a new command-line option `--debug-tileset` that enables `info` tileset messages.

File: client/tileset/tilespec.cpp [hunk 1]
```diff
@@ -1488,8 +1488,8 @@ static void tileset_add_layer(struct tileset *t, mapview_layer layer)
         t, t->city_offset, t->city_flag_offset, t->occupied_offset));
   } break;
   case LAYER_CITY2: {
-    t->layers.emplace_back(
-        std::make_unique<freeciv::layer_city_size>(t, t->city_size_offset));
+    t->layers.emplace_back(std::make_unique<freeciv::layer_city_size>(
```
Comment (jwrober) [inline client/tileset/tilespec.cpp:1491]: std library, need an include in header?

Comment (lmoureaux) [reply client/tileset/tilespec.cpp:1491]: It's only needed in the header if used there (which is the case). But `vector` is somehow magically included everywhere already, I'm not sure which header does that. If you want we can run `iwyu` in another PR...

Comment (jwrober) [reply client/tileset/tilespec.cpp:1491]: ok, i'll do a fresh compile on clang as a test as well

File: client/client_main.cpp [hunk 2]
```diff
@@ -374,7 +374,9 @@ int client_main(int argc, char *argv[])
        {{"S", "Sound"}, _("Read sound tags from FILE."), "FILE"},
        {{"m", "music"}, _("Read music tags from FILE."), "FILE"},
        {{"t", "tiles"}, _("Use data file FILE.tilespec for tiles."), "FILE"},
-       {{"w", "warnings"}, _("Warn about deprecated modpack constructs.")}});
+       {{"w", "warnings"}, _("Warn about deprecated modpack constructs.")},
+       {"debug-tileset", _("Prints tileset debugging information.")}});
```
Comment (jwrober) [inline client/client_main.cpp:378]: I don't see this in compiled code

```
$ build_2346/freeciv21-client --help
Encodings: Data=UTF-8, Local=en_US, Internal=UTF-8
Usage: build_2346/freeciv21-client [options] [url]

Options:
  -h, --help              Displays help on commandline options.
  --help-all              Displays help including Qt specific options.
  -v, --version           Displays version information.
  -A, --Announce <PROTO>  Announce game in LAN using protocol PROTO
                          (IPv4/IPv6/none).
  -a, --autoconnect       Skip connect dialog.
  -d, --debug <LEVEL>     Set debug log level
                          (fatal/critical/warning/info/debug).
  -F, --Fatal             Raise a signal on failed assertion.
  -f, --file <FILE>       Load saved game FILE.
  -H, --Hackless          Do not request hack access to local, but not spawned,
                          server.
  -l, --log <FILE>        Use FILE as logfile (spawned server also uses this).
  -M, --Meta <HOST>       Connect to the metaserver at HOST.
  -n, --name <NAME>       Use NAME as username on server.
  -p, --port <PORT>       Connect to server port PORT (usually with -a).
  -P, --Plugin <PLUGIN>   Use PLUGIN for sound output.
  -r, --read <FILE>       Read startup script FILE (for spawned server only).
  -s, --server <HOST>     Connect to the server at HOST (usually with -a).
  -S, --Sound <FILE>      Read sound tags from FILE.
  -m, --music <FILE>      Read music tags from FILE.
  -t, --tiles <FILE>      Use data file FILE.tilespec for tiles.
  -w, --warnings          Warn about deprecated modpack constructs.

Arguments:
  url                     Server information in URL format.
```

Comment (lmoureaux) [reply client/client_main.cpp:378]: I do...
```
$ build/freeciv21-client --help
Encodings: Data=UTF-8, Local=fr_BE, Internal=UTF-8
Utilisation : build/freeciv21-client [options] [url]

Options :
  -h, --help              Displays help on commandline options.
  --help-all              Displays help including Qt specific options.
  -v, --version           Afficher les informations de version.
  -A, --Announce <PROTO>  Announce game in LAN using protocol PROTO
                          (IPv4/IPv6/none).
  -a, --autoconnect       Skip connect dialog.
  -d, --debug <LEVEL>     Set debug log level
                          (fatal/critical/warning/info/debug).
  -F, --Fatal             Raise a signal on failed assertion.
  -f, --file <FILE>       Load saved game FILE.
  -H, --Hackless          Do not request hack access to local, but not spawned,
                          server.
  -l, --log <FILE>        Use FILE as logfile (spawned server also uses this).
  -M, --Meta <HOST>       Connect to the metaserver at HOST.
  -n, --name <NAME>       Use NAME as username on server.
  -p, --port <PORT>       Connect to server port PORT (usually with -a).
  -P, --Plugin <PLUGIN>   Use PLUGIN for sound output.
  -r, --read <FILE>       Read startup script FILE (for spawned server only).
  -s, --server <HOST>     Connect to the server at HOST (usually with -a).
  -S, --Sound <FILE>      Read sound tags from FILE.
  -m, --music <FILE>      Read music tags from FILE.
  -t, --tiles <FILE>      Use data file FILE.tilespec for tiles.
  -w, --warnings          Warn about deprecated modpack constructs.
  --debug-tileset         Prints tileset debugging information.

Arguments :
  url                     Server information in URL format.
  ```

Comment (jwrober) [reply client/client_main.cpp:378]: I started a fresh test from scratch. hm...

Comment (jwrober) [reply client/client_main.cpp:378]: I should note that the start up noise is reduced, so some of the changes are working.

**Review by jwrober — APPROVED** [review]
Requested changes look great.

**Pull request merged**